### PR TITLE
Codebreaker game

### PR DESCRIPTION
**Pull request by mmrshk**

### CODE REVIEW

File: lib/entities/game.rb [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    return message(:have_no_hints_message) unless @hint_avaliable == true
+
+    number = @process.hint_process(@code)
+    @hints_used += 1
+    puts "Digit #{number} on place #{@code.index(number) + 1} "
+    @hint_avaliable = false if @hints_used == @hints
+  end
+
+  def choice_process(command)
+    commands_in_game(command) unless command =~ /^[1-6]{4}$/
```
Comment (mmrshk) [inline lib/entities/game.rb]: between, without regular

File: Gemfile [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+source 'https://rubygems.org'
+
+git_source(:github) { |repo_name| "https://github.com/#{repo_name}" }
+
+#gem 'faker', :git => 'https://github.com/stympy/faker.git', :branch => 'master'
+
+#gem 'fasterer', '~> 0.4.1'
+
+#gem 'rubocop', '~> 0.60.0'
+
```
Comment (mmrshk) [inline Gemfile]: why it commented?

File: index.rb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+# frozen_string_literal: true
+
+require_relative 'autoload.rb'
+
+menu = Menu.new
+puts menu.game_menu
```
Comment (mmrshk) [inline index.rb]: delete puts

File: spec/codebreaker/processor_spec.rb [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+      [
+        [5, 6, 5, 6], [' ', ' ', ' ', ' ']
+      ]
+    ].each do |value|
+      it "tests that #{value.first} equals to #{value.last}" do
+        expect(subject.place_match(code, value.first)).to eq value.last
+      end
+    end
+  end
+
```
Comment (mmrshk) [inline spec/codebreaker/processor_spec.rb]: rethink

File: lib/entities/menu.rb [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+  def save_result
+    choice_save_process(ask(:save_results_message))
+  end
+
+  def registration
+    @name = ask(:registration)
+    registration unless name_valid?
+  end
+
+  def name_valid?
```
Comment (dzemlianoi-double) [inline lib/entities/menu.rb]: try to rethink

File: lib/entities/menu.rb [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    true
+  end
+
+  def level_choice
+    @level = ask(:hard_level)
+    choose_level(@level)
+  end
+
+  def secret_code
+    until lose?
```
Comment (dzemlianoi-double) [inline lib/entities/menu.rb]: try to unify

File: lib/entities/menu.rb [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+
+  def lose?
+    return if @game.attempts.positive?
+
+    @renderer.lost_game_message(@game.code)
+    false
+  end
+
+  def lost_hints
+    return @renderer.no_hints_message? if @game.hints.empty?
```
Comment (dzemlianoi-double) [inline lib/entities/menu.rb]: hints_spent?

File: lib/entities/menu.rb [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+  end
+
+  def win(result)
+    return if result != WIN
+
+    @renderer.win_game_message
+    true
+  end
+
+  def save_game_result
```
Comment (dzemlianoi-double) [inline lib/entities/menu.rb]: try to move it to separate class/method, maybe to_h of Game class

File: lib/entities/menu.rb [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+
+  def game_operations(command_name)
+    case command_name.to_sym
+    when HINT then lost_hints
+    when EXIT then game_menu
+    else valid_command?(command_name)
+    end
+  end
+
+  def valid_command?(command)
```
Comment (dzemlianoi-double) [inline lib/entities/menu.rb]: try to rethink

File: lib/entities/menu.rb [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+    if command.to_i.zero?
+      @renderer.command_error
+      secret_code
+    elsif check_command_range(command).nil?
+      @renderer.command_int_error
+      secret_code
+    end
+  end
+
+  def choose_level(level)
```
Comment (dzemlianoi-double) [inline lib/entities/menu.rb]: rethink

File: lib/locales/en.yml [hunk 11]
```diff
@@ -0,0 +1,49 @@
+en:
+  start_message: "Welcome to CodeBreaker"
+  choice_options: "Choose option:
+  1. start
```
Comment (dzemlianoi-double) [inline lib/locales/en.yml]: dynamic

File: lib/modules/validator.rb [hunk 12]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+module Validator
+  def check_name_emptyness(name)
+    return true unless name.empty?
```
Comment (dzemlianoi-double) [inline lib/modules/validator.rb]: !name.empty?

File: lib/entities/menu.rb [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+    choice_code_process
+    renderer.round_message
+    game.decrease_attempts!
+  end
+
+  def choice_code_process
+    case guess
+    when Game::HINT then hint_process!
+    when COMMANDS[:exit] then game_menu
+    else handle_command(guess)
```
Comment (6aBblKuH) [inline lib/entities/menu.rb]: remove arg

File: lib/entities/menu.rb [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+    end
+  end
+
+  def handle_command(command)
+    return p game.start_process(command) if check_command_range(command)
```
Comment (6aBblKuH) [inline lib/entities/menu.rb]: use ```guess``` instead ```command```

File: lib/entities/menu.rb [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+  def ask(phrase_key = nil, options = {})
+    renderer.message(phrase_key, options) if phrase_key
+    gets.chomp
+  end
+
+  def save_result
+    storage.save_game_result(game.to_h(@name)) if ask(:save_results_message) == CHOOSE_COMMANDS[:yes]
+  end
+
+  def registration
```
Comment (6aBblKuH) [inline lib/entities/menu.rb]: rename

File: lib/entities/game.rb [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+  HINT = 'hint'
+
+  attr_reader :attempts, :hints, :code
+
+  def initialize
+    @process = Processor.new
+  end
+
+  def generate(difficulty)
+    @level = difficulty
```
Comment (6aBblKuH) [inline lib/entities/game.rb]: rename

File: lib/entities/game.rb [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+    @code = generate_secret_code
+    @hints = @code.sample(DIFFICULTIES[difficulty][:hints])
```
Comment (6aBblKuH) [inline lib/entities/game.rb]: use ```@level```

File: lib/entities/menu.rb [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+  end
+
+  def start
+    @name = registration
+    level_choice
+    game_process
+  end
+
+  def stats
+    users_data = storage.load
```
Comment (6aBblKuH) [inline lib/entities/menu.rb]: rename

File: lib/entities/menu.rb [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+
+  def name_valid?(name)
+    !check_emptyness(name) && check_length(name, MIN_SIZE_VALUE, MAX_SIZE_VALUE)
+  end
+
+  def level_choice
+    loop do
+      level = ask(:hard_level, levels: Game::DIFFICULTIES.keys.join(' | '))
+
+      return generate_game(level.to_sym) if Game::DIFFICULTIES[level.to_sym]
```
Comment (6aBblKuH) [inline lib/entities/menu.rb]: ```generate_game(Game::DIFFICULTIES[level.to_sym])```

File: lib/entities/menu.rb [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+    when COMMANDS[:exit] then game_menu
+    else handle_command(guess)
+    end
+  end
+
+  def handle_command(command)
+    return p game.start_process(command) if check_command_range(command)
+
+    renderer.command_error
+    game_process
```
Comment (6aBblKuH) [inline lib/entities/menu.rb]: remove

File: lib/entities/menu.rb (hunk 14 above)
Comment (6aBblKuH) [reply lib/entities/menu.rb]: return error unless valid

File: lib/entities/menu.rb [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+    save_result
+    game_menu
+  end
+
+  def handle_lose
+    renderer.lost_game_message(game.code)
+    game_menu
+  end
+
+  def hint_process!
```
Comment (6aBblKuH) [inline lib/entities/menu.rb]: ```hint_process```

File: lib/entities/statistics.rb [hunk 22]
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+class Statistics
+  def stats(list)
+    hash = list.group_by { |score| score[:difficulty] }
```
Comment (6aBblKuH) [inline lib/entities/statistics.rb]: rename to difficulties

File: lib/entities/statistics.rb [hunk 23]
```diff
@@ -0,0 +1,16 @@
...
+    %w[hell medium easy].reduce([]) do |previous, current|
```
Comment (6aBblKuH) [inline lib/entities/statistics.rb]: ```do |sorted_difficulties, difficulty_name|```

File: lib/entities/statistics.rb [hunk 24]
```diff
@@ -0,0 +1,16 @@
...
+      hash[current] ? previous + stats_sort(hash[current]) : previous
+    end
+  end
+
+  private
+
+  def stats_sort(players_array)
```
Comment (6aBblKuH) [inline lib/entities/statistics.rb]: rename to ```scores```

File: lib/entities/statistics.rb [hunk 25]
```diff
@@ -0,0 +1,16 @@
...
+    players_array.sort_by! { |k| [k[:attempts_used], k[:hints_used]] }.reverse
```
Comment (6aBblKuH) [inline lib/entities/statistics.rb]: rename ```k```

**Pull request merged**